setuptools/config/pyprojecttoml.py

```diff
@@ -365,6 +374,35 @@ def _obtain_classifiers(self, dist: "Distribution"):
                 return value.splitlines()
         return None
 
+    def _obtain_dependencies(self, dist: "Distribution"):
+        if "dependencies" in self.dynamic:
+            value = self._obtain(dist, "dependencies", {})
+            if value:
+                return _parse_requirements_list(value)
+        return None
+
+    def _obtain_optional_dependencies(self, dist: "Distribution"):
+        if "optional-dependencies" in self.dynamic:
+            optional_dependencies_map = self.dynamic_cfg["optional-dependencies"]
+            assert isinstance(optional_dependencies_map, dict)
+            return {
+                group: _parse_requirements_list(self._expand_directive(
+                    f"tool.setuptools.dynamic.optional-dependencies.{group}",
+                    directive,
+                    {},
+                ))
+                for group, directive in optional_dependencies_map.items()
+            }
```

Comment:
Can we also have the `self._ensure_previously_set(dist, field)` before the last return in this function?

Since it does not use `self._obtain` directly, there is no contingency for the case we `optional-dependencies` is listed as `dynamic` but not provided in `tool.setuptools.dynamic`...

PEP 621 requires the following:
> Build back-ends MUST raise an error if the metadata specifies a field in dynamic but the build back-end was unable to provide the data for it.